rust/sdk-processor/src/steps/parquet_default_processor/size_buffer.rs

```diff
@@ -0,0 +1,260 @@
+use crate::{
+    parquet_processors::{ParquetTypeEnum, ParquetTypeStructs},
+    steps::parquet_default_processor::gcs_handler::Uploadable,
+};
+use anyhow::Result;
+use aptos_indexer_processor_sdk::{
+    traits::{
+        pollable_async_step::PollableAsyncRunType, NamedStep, PollableAsyncStep, Processable,
+    },
+    types::transaction_context::{TransactionContext, TransactionMetadata},
+    utils::errors::ProcessorError,
+};
+use async_trait::async_trait;
+use std::{collections::HashMap, time::Duration};
+use tracing::debug;
+
+struct ParquetBuffer {
+    pub buffer: ParquetTypeStructs,
+    pub buffer_size_bytes: usize,
+    current_batch_metadata: TransactionMetadata,
+}
+
+impl ParquetBuffer {
+    fn new(parquet_type: &ParquetTypeEnum) -> Self {
+        Self {
+            buffer: ParquetTypeStructs::default_for_type(parquet_type),
+            buffer_size_bytes: 0,
+            current_batch_metadata: TransactionMetadata::default(),
+        }
+    }
+
+    pub fn update_current_batch_metadata(
+        &mut self,
+        cur_batch_metadata: &TransactionMetadata,
+        uploaded: bool,
+    ) {
+        let buffer_metadata = &mut self.current_batch_metadata;
+        buffer_metadata.end_version = cur_batch_metadata.end_version;
+        buffer_metadata.total_size_in_bytes = cur_batch_metadata.total_size_in_bytes;
+        buffer_metadata.end_transaction_timestamp =
+            cur_batch_metadata.end_transaction_timestamp.clone();
+        // if it wasn't uploaded -> we update only end_version, size, and last timestamp
+        if uploaded {
+            buffer_metadata.start_version = cur_batch_metadata.start_version;
+            buffer_metadata.start_transaction_timestamp =
+                cur_batch_metadata.start_transaction_timestamp.clone();
+        }
+    }
+}
+
+pub struct SizeBufferStep<U>
+where
+    U: Uploadable + Send + 'static + Sync,
+{
+    internal_buffers: HashMap<ParquetTypeEnum, ParquetBuffer>,
+    pub internal_buffer_size_bytes: usize,
+    pub poll_interval: Duration,
+    pub buffer_uploader: U,
+    pub buffer_max_size: usize,
+}
+
+impl<U> SizeBufferStep<U>
```

Comment:
Nice job for first version! It'd also be great to have some unit tests around this step.

Reply:
will follow-up with some unit tests